Doc: seal migration update (#8724)
hashicorp/vault: website/pages/docs/concepts/seal.mdx

currently supported when using Vault Enterprise. We plan to support this
officially in a future release.

### Migration post Vault 1.4.0

These steps are common for seal migrations between any supported kinds and for
any storage backend.

1. Take the standby nodes down, one node at a time and update the [seal
configuration](/docs/configuration/seal). If the migration is from Shamir seal
to KMS seal, add the desired new KMS seal block to the config. If the migration
is from KMS seal to Shamir seal, add `disabled = "true"` to the old seal block.
If the migration is from KMS seal to another KMS seal, add `disabled = "true"`
to the old seal block and add the desired new KMS seal block.

2. Bring the standby nodes back up one at a time and run the unseal command on
each by supplying the `-migrate` flag. Supply Shamir unseal keys if old seal was
Shamir, which will be migrated as the recovery keys for the KMS seal. Supply
recovery keys if the old seal is one of KMS seals, which will be migrated as the
recovery keys of the new KMS seal or as Shamir unseal keys if the new seal is
Shamir.

3. Stop the active node. One of the standby nodes will become the active node

> and performs the migration. If possible, monitor the server log in the active

Reviewer: ```suggestion
and perform the migration. If possible, monitor the server log in the active
```

node to witness the completion of the seal migration process.

4. Seal migration is now completed. Update the config of the old active node
(that is still down) to use the new seal blocks (completely unaware of the old

> seal type) and bring it. It will be auto-unsealed if the new seal is one of the

Reviewer: ```suggestion
seal type) and bring it up. It will be auto-unsealed if the new seal is one of the
```

decrypt the key but will not unseal Vault. When you bring your server back up,
run the unseal process with the `-migrate` flag and use the Recovery Keys to
perform the migration. All unseal commands must specify the `-migrate` flag.
Once the required threshold of recovery keys are entered, the recovery keys will
be migrated to be used as unseal keys.

#### Migration From Auto Unseal to Auto Unseal

~> **NOTE**: Migration between same Auto Unseal types is not currently
supported. We plan to support this officially in a future release.

To migrate from Auto Unseal to a different Auto Unseal configuration, take your
server cluster offline and update the existing [seal
configuration](/docs/configuration/seal) and add `disabled = "true"` to the seal
block.  Then add another seal block to describe the new seal.

When you bring your server back up, run the unseal process with the `-migrate`
flag and use the Recovery Keys to perform the migration. All unseal commands
must specify the `-migrate` flag. Once the required threshold of recovery keys
are entered, the recovery keys will be kept and used as recovery keys in the new
seal.

> #### Migration with Integrated Storage

Reviewer: Given the header nesting, I read this as still Integrated Storage pre 1.4, and that integrated storage post 1.4 would use the other set of instructions - is that accurate?

Author: Yes. That is correct.

Vault CLI, this is performed by using the `-target=recovery` flag to `vault operator rekey`.

## Seal Migration

The seal can be migrated from Shamir Seal to KMS Seal, KMS Seal to Shamir Seal,
and KMS Seal to another KMS Seal.

~> **NOTE**: Seal migration process cannot be performed without downtime. Due to
the technical underpinnings of the seal implementations, it is at this point not
possible to perform seal migration without briefly bringing the whole cluster
down. We understand that it can be hard for many deployments to face downtime,
but we believe that switching seals is a rare event and hence we hope for the
downtime to be considered as an acceptable trade-off.

~> **NOTE**: Seal migration operation will require both old and new seals to be
available during the migration. For example, migration from KMS seal to Shamir
seal will require that the KMS key be accessible during the migration.

~> **NOTE**: Seal migration from KMS seal to Shamir seal is not currently
supported when using Vault Enterprise. We plan to support this officially in a
future release.

~> **NOTE**: Seal migration from KMS seal to KMS seal of same kind is not

> currently supported when using Vault Enterprise. We plan to support this

Reviewer 2: I think this is unsupported in OSS too, right?

Author: Correct. Updated.

officially in a future release.

### Migration post Vault 1.4.0

These steps are common for seal migrations between any supported kinds and for
any storage backend.

1. Take the standby nodes down, one node at a time and update the [seal
configuration](/docs/configuration/seal). If the migration is from Shamir seal
to KMS seal, add the desired new KMS seal block to the config. If the migration
is from KMS seal to Shamir seal, add `disabled = "true"` to the old seal block.
If the migration is from KMS seal to another KMS seal, add `disabled = "true"`
to the old seal block and add the desired new KMS seal block.

2. Bring the standby nodes back up one at a time and run the unseal command on
each by supplying the `-migrate` flag. Supply Shamir unseal keys if old seal was
Shamir, which will be migrated as the recovery keys for the KMS seal. Supply
recovery keys if the old seal is one of KMS seals, which will be migrated as the
recovery keys of the new KMS seal or as Shamir unseal keys if the new seal is

> Shamir.

Reviewer 2: We should make it clear that step 1 & 2 should be done per node. e.g. Do step 1&2 on standby 1 then Do step 1&2 on standby 2, etc. Otherwise if you do step 1 on standby 1&2 then step 2 on standby 1&2 you've likely broken raft quorum.

Author: Merged step 1&2 as a single block of action and updated step 2 to have step 1 in a modular way. Let me know what you think of that.

3. Stop the active node. One of the standby nodes will become the active node
and perform the migration. If possible, monitor the server log in the active
node to witness the completion of the seal migration process.

4. Seal migration is now completed. Update the config of the old active node
(that is still down) to use the new seal blocks (completely unaware of the old

> seal type) and bring it up. It will be auto-unsealed if the new seal is one of the

Reviewer 2: If using raft storage this isn't true. The old KMS will be needed to unseal the node until the new keys are  replicated. We should make sure all nodes have access to both KMSs for a short while. Then after values are done being seal re-wrapped and all data has been replicated to raft nodes we can remove the old configs.

Author: I think we can achieve quorum without the active node up since raft can sustain loss of one node. So, I've updated step 3 to cover this case. Let me know if that will help resolve this issue.

KMS seals or will require unseal keys if the new seal is Shamir.

5. At this point, config files of all the nodes can be updated to only have the
new seal information. Standby nodes can be restarted right away and the active

> node can be restarted upon a leadership change.

Reviewer 2: Should probably make sure all sealwrapped data is finished re-wrapping before doing this step.

Author: I've updated step 3 to cover this. Let me know what you think of that.

### Migration pre 1.4

#### Migration From Shamir to Auto Unseal
To migrate from Shamir keys to Auto Unseal, take your server cluster offline and
update the [seal configuration](/docs/configuration/seal) with the appropriate
seal configuration. Bring your server back up and leave the rest of the nodes
offline if using multi-server mode, then run the unseal process with the
`-migrate` flag and bring the rest of the cluster online.

All unseal commands must specify the `-migrate` flag. Once the required
threshold of unseal keys are entered, unseal keys will be migrated to recovery
keys.

``` $ vault operator unseal -migrate ```

#### Migration From Auto Unseal to Shamir

> ~> **NOTE**: Migration to Shamir seal is not currently supported when using

Reviewer 2: Seems like this is now duplicated above and here

Author: Removed. Thanks!

The seal can be migrated from Shamir Seal to KMS Seal, KMS Seal to Shamir Seal,
and KMS Seal to another KMS Seal.

~> **NOTE**: Seal migration process cannot be performed without downtime. Due to
the intricate technical underpinnings of the seal feature, it is at this point
not possible to perform seal migration without bringing the whole cluster down.
We understand that it can be painful for many users to bring the full cluster
down. But, we believe that switching seals is a very rare event and hence hope
for the downtime to be considered as an acceptable trade-off.

~> **NOTE**: Seal migration operation will require both seals to be available
during the migration. For example, a migration from a KMS seal to Shamir will
require that the KMS key be accessible during the migration.

~> **NOTE**: Seal migration from KMS seal to Shamir seal is not currently
supported when using Vault Enterprise. We plan to support this officially in a
future release.

~> **NOTE**: Seal migration from KMS seal to KMS seal of same kind is not
currently supported when using Vault Enterprise. We plan to support this
officially in a future release.

### Migration From Shamir Seal to KMS Seal

> #### When **not** using Integrated Storage

Reviewer 2: I don't understand why we cannot use one set of instructions for any storage type. Since only the active node does the migration can't we follow the same steps as integrated storage for any storage backend. I think the following steps will work for any storage:

1. Take Standby nodes down and update their configs.
2. Start the Standby nodes and unseal them with -migrate
3. Stop the active node.
4. One of the standby nodes will become active and preform the migration.
5. Depending on the migration type update the config on the old active node and start it.

This flow also minimizes downtime to the failover time + the time it takes to migrate. Which is better than having to stop all nodes at once.

Author: The primary reason for this approach is cleanup.

For storage backends that are not using raft, there is really no need to setup all the nodes in the cluster for migration. The first/active node fully completes the operation and standby nodes can join the cluster with no knowledge of migration (no -migrate flag or no old seal blocks required). I thought this could be easier to reason and work with. Seal migration requires the config file to be cleaned up after the migration is complete (removal of old seal blocks). Having one node perform the migration means that there is one node's config that needs cleanup. Downtime is anyways inevitable. I thought that it might be better to work with less moving parts than probably saving a few seconds of downtime.

For raft storage, migration can be performed by quorum number of nodes. Then cleanup of all the quorum number of nodes' config file through failover/update approach. After this, addition of the rest of the peer nodes will be with no knowledge of migration.

Author: I am okay to go either ways. Just wanted to let you know the thought process.

Reviewer 2: > Downtime is anyways inevitable. I thought that it might be better to work with less moving parts than probably saving a few seconds of downtime.

I'm not sure this is true, with the steps I outlined above the only disruption you'd have is during the failover from active to standby. This is much less than shutting all nodes down, updating a config file, and starting them up. 

> For storage backends that are not using raft, there is really no need to setup all the nodes in the cluster for migration.

It depends on what you're migrating to/from. You may want to include the migration seal on all nodes incase you need to unseal-wrap a lot of data. 

Plus I think it's simpler to have one set of instructions than to recommend different practices based on your storage backend. 

I also would like to see a pre-1.4 set of steps and a post-1.4 set of steps since the approach changed quite a bit in 1.4.

Author: @briankassouf Let me know what you think of the updates.

For a list of examples and supported providers, please see the
[seal documentation](/docs/configuration/seal).

## Recovery Key Rekeying

> During KMS Seal initialization process, a set of Shamir keys called Recovery Keys are

Reviewer: ```suggestion
During the KMS Seal initialization process, a set of Shamir keys called Recovery Keys are
```

> generated which are used for operations that still require a quorum of users.

Reviewer: ```suggestion
generated which are used for operations that still require a quorum of keyholders.
```

Recovery Keys can be rekeyed to change the number of shares or thresholds. When using the
Vault CLI, this is performed by using the `-target=recovery` flag to `vault operator rekey`.

## Seal Migration

> The seal can be migrated from Shamir Seal to KMS Seal, KMS Seal to Shamir Seal

Reviewer: ```suggestion
The seal can be migrated from Shamir Seal to KMS Seal, KMS Seal to Shamir Seal,
```

and KMS Seal to another KMS Seal.

~> **NOTE**: The seal migration process cannot be performed without downtime.

> Due to the intricate technical underpinnings of the seal feature, it is at this

Reviewer: I would delete "at this point" but couldn't figure out how to do a multiline suggestion.

Author: Done.

When clicking the `+` button, you hold and drag it down first to select multiple lines and then click it to do a multi line comment.

point not possible to perform seal migration without bringing the whole cluster
down. We understand that it can be painful for many users to bring the full
cluster down. But, we believe that switching seals is a very rare event and
hence hope for the downtime to be considered as an acceptable trade-off.

~> **NOTE**: The migration operation will require both seals to be available
during the migration. For example, a migration from a KMS seal to Shamir will
require that the KMS key be accessible during the migration.

### Migration From Shamir Seal to KMS Seal

#### When **not** using Integrated Storage

1. Take your server cluster offline.

2. Add the desired KMS [seal block](/docs/configuration/seal) to the
configuration file of one of the nodes.

> 3. Bring this one node back up and run the unseal command by supplying the

Reviewer: A link here to the unseal command would be nice, I think.

`-migrate` flag for each unseal key. This node will be the active node and
completes the seal migration process. The Shamir unseal keys supplied here will
be migrated as the recovery keys for the KMS seal.

> 4. Standby nodes can join the active node's cluster using the new seal blocks.

Reviewer: I don't understand "using the new seal blocks" - what needs to happen on the standby nodes? Or can you just re-start them and they will join because of the config block on the original node?

Note that by the time standby nodes join the cluster, the migration process is
already complete. Hence, the standby nodes will be auto-unsealed through the KMS
key and will not require an explicit unseal command.

#### When using Integrated Storage

Integrated Storage is implemented using the [Raft Consensus Algorithm][raft] and

> would require a quorum of nodes to be fully operational.

Reviewer: ```suggestion
requires a quorum of nodes to be fully operational.
```

1. Take your server cluster offline.

2. Add the desired KMS [seal block](/docs/configuration/seal) to the

> configuration files of quorum number of nodes in the cluster.

Reviewer: ```suggestion
configuration files of a quorum number of nodes in the cluster.
For example, if your cluster is 5 nodes, adding the configuration to 3 nodes would give you quorum.
```

I think this example is correct? Anyway, I think an example would be helpful.

> 3. Bring the quorum number of nodes back up. On each of these nodes, run the

Reviewer: Maybe break this into two steps - the first bringing the nodes back up, and the second running the unseal command on each of the nodes (unless the order is bring them back up and run the unseal command on each before bringing the next up, in which case, I think we need to be more explicit).

unseal command by supplying the `-migrate` flag for each unseal key. The Shamir
unseal keys supplied here will be migrated as the recovery keys for the KMS
seal. Raft protocol will then elect a leader node and this leader node will
complete the seal migration process. If a leadership change happens after the
seal migration is complete, the new leader will detect the fact and does not
repeat the seal migration.

> 4. Any new node joining to the cluster at this point will not require manual

Reviewer: ```suggestion
4. Any new node joining the cluster at this point will not require manual
```

running of the unseal command. The join will ensure auto-unsealing of the node

> through KMS key.

Reviewer: When you bring up the non-quorum nodes, do they need to re-join the cluster with the join command? If so, can you make that more obvious in this instruction? Ie: These nodes will still need to re-join the cluster using the join command (+link).

### Migration From KMS Seal to Shamir Seal

> ~> **NOTE**: Migration to Shamir seal is not currently supported when using

Reviewer: I think this note belongs under the top-level header - right after the first paragraph. We may want to add that KMS to same KMS is not currently supported.

Author: Done. Thanks for reminding about KMS to KMS of same kind. I had it in my mind to do it and it slipped through.

Vault Enterprise. We plan to support this officially in a future release.

#### When **not** using Integrated Storage

1. Take your server cluster offline.

> 2. Add `disabled = "true"` to the old KMS [seal block](/docs/configuration/seal)

Reviewer: Do we have an example on the linked page of where this should go in the block?

in the config file of one of the nodes. Shamir seal does not require an explicit
seal block.

3. Bring this one node back up and run the unseal command using the recovery
keys of the KMS seal along with supplying the `-migrate` flag for each recovery
key. This node will be the active node and completes the seal migration process.
The recovery keys supplied here will be migrated as the unseal keys for the
Shamir seal.

> 4. Standby nodes can join the active node's cluster without the KMS seal blocks.

Reviewer: I have the same question here that I did on the corresponding step earlier (basically, can you just re-start them).

Note that by the time standby nodes join the cluster, the migration process is
already complete. Hence, the standby nodes will not require the `-migrate` flag
to be unsealed.

#### When using Integrated Storage

Integrated Storage is implemented using the [Raft Consensus Algorithm][raft] and

> would require a quorum of nodes to be fully operational.

Reviewer: ```suggestion
requires a quorum of nodes to be fully operational.
```

1. Take your server cluster offline.

2. Add `disabled = "true"` to the old KMS [seal block](/docs/configuration/seal)

> to the configuration files of quorum number of nodes in the cluster. Shamir seal

Reviewer: ```suggestion
in the configuration files of a quorum number of nodes in the cluster. The Shamir seal
```

Reviewer: Also I think an example of the quorum number would be helpful here.

does not require an explicit seal block.

> 3. Bring the quorum number of nodes back up. On each of these nodes, run the

Reviewer: I have the same question about the order here (bring all up then migrate, or bring one up then migrate then repeat?).

unseal command using the recovery keys of the KMS seal along with supplying the
`-migrate` flag for each recovery key. The recovery keys supplied here will be
migrated as the unseal keys for the Shamir seal. Raft protocol will then elect a
leader node and this leader node will complete the seal migration process. If a
leadership change happens after the seal migration is complete, the new leader
will detect the fact and does not repeat the seal migration.

### Migration From KMS Seal to another KMS Seal

~> **NOTE**: Migration between same Auto Unseal types is not currently
supported. We plan to support this officially in a future release.

#### When **not** using Integrated Storage

1. Take your server cluster offline.

2. Add `disabled = "true"` to the old KMS [seal block](/docs/configuration/seal)
and add the desired new seal block in the config file of one of the nodes.

3. Bring this one node back up and run the unseal command using the Recovery

> Keys of the old KMS seal along with supplying the `-migrate` flat for each

Reviewer: ```suggestion
Keys of the old KMS seal along with supplying the `-migrate` flag for each
```

> recovery key. This node will be the active node and completes the seal migration

Reviewer: Can we standardize the capitalization as either Recovery Key or recovery key?

Author: Done.

process. The recovery keys of the old KMS seal supplied here will be migrated as
the recovery keys for the new KMS seal.

4. Standby nodes can join the active node's cluster without the old KMS seal
blocks. Note that by the time standby nodes join the cluster, the migration
process is already complete. Hence, the standby nodes will be auto-unsealed
through the new KMS key.

#### When using Integrated Storage

Integrated Storage is implemented using the [Raft Consensus Algorithm][raft] and

> would require a quorum of nodes to be fully operational.

Reviewer: ```suggestion
requires a quorum of nodes to be fully operational.
```

1. Take your server cluster offline.

2. Add `disabled = "true"` to the old KMS [seal block](/docs/configuration/seal)

> and add the desired new seal block in the config file of quorum number of nodes

Reviewer: ```suggestion
and add the desired new seal block in the config file of a quorum number of nodes
```

in the cluster.

3. Bring the quorum number of nodes back up. On each of these nodes, run the
unseal command using the recovery keys of the old KMS seal along with supplying

> the `-migrate` flat for each recovery key. The recovery keys of the old KMS seal

Reviewer: ```suggestion
the `-migrate` flag for each recovery key. The recovery keys of the old KMS seal
```

> node will perform the migration. The migrated information will be replicated to

Reviewer: ```suggestion
node that will perform the migration. The migrated information will be replicated to
```
